src/hardware/mixer.cpp

```diff
@@ -1324,13 +1325,44 @@ std::unique_ptr<Program> MIXER_ProgramCreate() {
 	return ProgramCreate<MIXER>();
 }
 
-void MIXER_Init(Section* sec) {
+void MIXER_SetState(const MixerState requested)
+{
+	assert(requested != MixerState::Uninitialized);
+
+	// alias
+	auto &current = mixer.state;
+
+	if (current == requested)
+		return;
+
+	// Off-state requested?
+	if ((requested == MixerState::NoSound || requested == MixerState::Mute) ||
+	    (requested == MixerState::Off && current != MixerState::Mute)) {
+		TIMER_DelTickHandler(MIXER_Mix);
+		TIMER_DelTickHandler(MIXER_Mix_NoSound);
+		TIMER_AddTickHandler(MIXER_Mix_NoSound);
+		SDL_PauseAudioDevice(mixer.sdldevice, 1);
+		current = requested;
+	}
+
+	// On-state requested?
+	else if (requested == MixerState::On || (requested == MixerState::FocusResume &&
+	                                         current != MixerState::Mute)) {
+		TIMER_DelTickHandler(MIXER_Mix);
```

Comment:
My view is that the `FocusResume` and `Off` states confuse the intent here; I've spent a fair bit of time jumping back and forth betwen the different places until I figured out why they're needed (in fact, they're not necessary).

I suggest to get rid of them as `MIXER_SetState` should not care about the _reason_ why a mute or unmute was requested. In fact, I don't really see why we should make the distinction between a simple unmute request (`MixerState::On`) and a "resume" request `(MixerState::FocusResume`). So when the window becomes focused, we'll just simply send a `MixerState::On` command to `MIXER_SetState` and that's about it. I don't see the value in making this more complicated than it has to be.

Reply:
I've had another look and I think what confuses things here (for me at least) is the conflation of state vs request; it's present even in your wording:

```
Off, // A system-controlled off-state, might be via pause or loss of focus.
Mute, // A user-controlled off-state via the hotkey action.
On, // A system-controlled on-state, might be via unpause or gain of focus.
FocusResume, // A request that, if not muted, will lead to the on-state.
```

I'm not saying to over-design it and have separate states and requests, with lots of duplication between them, but maybe just restrict it to actual states, then the caller can request a new state that gets stored as the current state, not send in a state that is actually a request but then gets transformed into some other actual state, using the same enum... To me that's confusing as hell.

Reply:
Suggestions rolled into the latest implementation (the orthogonal `FocusResume` has been removed).